presto-main/src/main/java/com/facebook/presto/cost/AggregationStatsRule.java

```diff
@@ -76,27 +95,50 @@ public static PlanNodeStatsEstimate groupBy(PlanNodeStatsEstimate sourceStats, C
             }));
         }
 
-        double rowsCount = 1;
+        double rowsCount = SINGLE_ROW_COUNT;
         for (VariableReferenceExpression groupByVariable : groupByVariables) {
             VariableStatsEstimate symbolStatistics = sourceStats.getVariableStatistics(groupByVariable);
             int nullRow = (symbolStatistics.getNullsFraction() == 0.0) ? 0 : 1;
             rowsCount *= symbolStatistics.getDistinctValuesCount() + nullRow;
         }
-        result.setOutputRowCount(min(rowsCount, sourceStats.getOutputRowCount()));
+        // groupBy key may be a derived expression which may not have associated stats hence return NaN.
+        // In such case, use source stats to bound the output row count
+        if (!isNaN(rowsCount)) {
+            result.setOutputRowCount(min(rowsCount, sourceStats.getOutputRowCount()));
+        }
+        else {
+            result.setOutputRowCount(sourceStats.getOutputRowCount());
+        }
 
         for (Map.Entry<VariableReferenceExpression, Aggregation> aggregationEntry : aggregations.entrySet()) {
-            result.addVariableStatistics(aggregationEntry.getKey(), estimateAggregationStats(aggregationEntry.getValue(), sourceStats));
+            result.addVariableStatistics(aggregationEntry.getKey(), estimateAggregationStats(aggregationEntry.getValue(), sourceStats, rowsCount));
+        }
+
+        if (isAggregateToSingleRow) {
+            result.setOutputRowCount(SINGLE_ROW_COUNT);
         }
 
         return result.build();
     }
 
-    private static VariableStatsEstimate estimateAggregationStats(Aggregation aggregation, PlanNodeStatsEstimate sourceStats)
+    private static VariableStatsEstimate estimateAggregationStats(Aggregation aggregation, PlanNodeStatsEstimate sourceStats, double estimatedOutputRowCount)
     {
         requireNonNull(aggregation, "aggregation is null");
         requireNonNull(sourceStats, "sourceStats is null");
+        requireNonNull(estimatedOutputRowCount, "estimatedOutputRowCount is null");
+
+        double width;
+
+        Type type = aggregation.getCall().getType();
+        if (type instanceof FixedWidthType) {
+            width = ((FixedWidthType) type).getFixedSize();
+        }
+        else {
+            width = DEFAULT_DATA_SIZE_PER_COLUMN;
+        }
 
-        // TODO implement simple aggregations like: min, max, count, sum
-        return VariableStatsEstimate.unknown();
+        // Double.MIN_VALUE is a constant holding the smallest positive nonzero value of type double.
+        // Put a -(minus sign) to become true negative smallest value to make correct cardinality estimate for a filter range expression to liberal comparison
+        return new VariableStatsEstimate(-Double.MIN_VALUE, Double.MAX_VALUE, 0, width, estimatedOutputRowCount);
```

Comment:
1. use NEGATIVE_INFINITY and POSITIVE_INFINITY to represent the full range.  
2. aggregations CAN return null on null input, so if the source stats are 100% null and the function is not called on null input, then the nulls fraction here will also be 1.0 (not sure if there are any aggs that return null on mix of null and non-null input, but should be okay to skip this case). 
3. it's not at all guaranteed that the aggregation columns will be distinct (for the stats estimates i guess it's ok as a heuristic, but NaN would be more accurate. It definitely needs a comment) As an example, if you did `SELECT count(*) from nation GROUP BY regionkey`, every row would have a value of 5.
4. as a further optimization for min/max/count that are only one row, we can  get the min, max or count of the source stats.

Reply:
#1. fixed
#2. if all input are null, isn't the aggregate of all nulls be a single null?
![image](https://github.com/prestodb/presto/assets/32091981/33867dac-57cf-430f-8589-165d6f6fd382)
#3 yes if agg columns are not all distinct, the estimate can only be heuristic in such case to return a max row count possible so that cost-based decision can avoid having NaN case 
![image](https://github.com/prestodb/presto/assets/32091981/b9a788b2-9cb4-41ac-ba6e-590a983145ca)
#4 if there's no global grouping key yes we can optimize further for min/max/count that only returns a single row output

Reply:
1. agree 
2. Aggregations can return NULL. But for most queries this is not the case and I do not see how we could understand/leverage that in the optimizer.
3. Returning NaN may be more "accurate" but isn't that what leads to the problem we are trying to solve? Ultimately almost all estimates used in the optimizer are "guesses", e.g. Presto uses 0.9 as the selectivity for some single table predicates. Not having some reasonable estimate for the NDV of an aggregate column leads to very bad plans even for simple cases, e.g. "explain select 1 from t1, t2 where a1=a2 and b1 = (select max(b3) from t3 group by a3)". Rewriting a query to get the desired join order is not an option in the vast majority of environments.
4. agree

Reply:
3. yeah, i think the heuristic is okay, but it would be good to have a small comment noting that it's a heuristic.  I feel like all these stats rules are a combo of things that are definitionally true of the operator (e.g ungrouped aggregation only returns a single row) and some heuristics that we expect to be true more often than not.  I don't think we've been very consistent about doing this throughout the stats rules, but for me it's helpful to call out when/why we are using a heuristic so it's easier to understand where the stats estimates come from.

Reply:
2. @fgwang7w your example is null grouping keys -  I mean the aggregation inputs are null: 
```
presto:di> with x as (SELECT CAST (null as INTEGER) a, 1 as b) SELECT count(a), sum(a), count(*), b from x group by b;
 _col0 | _col1 | _col2 | b 
-------+-------+-------+---
     0 | NULL  |     1 | 1 
```

I was suggesting that since we have the null fraction for a column and info for each function whether it returns null on null input, we could use that to be more precise for all null input if we wanted (for an additional improvement beyond what's covered in this PR)

Reply:
Thank you for the hints. I refactor the code a little bit, please help review.